[GLib] Need API for asynchronously handling WebKitDownload::decide-destination

mcatanzaro opened the pull request:
#### 878af52cf35811d107d6bb6f38531ce3b73f211d
<pre>
[GLib] Need API for asynchronously handling WebKitDownload::decide-destination
<a href="https://bugs.webkit.org/show_bug.cgi?id=238748">https://bugs.webkit.org/show_bug.cgi?id=238748</a>

Reviewed by Carlos Garcia Campos.

Currently WebKitDownload::decide-destination is emitted when WebKit
wants to know the filesystem location to download the file to. The
application is expected to display a file chooser using
gtk_dialog_run(), then return the result to WebKit using
webkit_download_set_destination(). In GTK 4, gtk_dialog_run() no longer
exists. The only way to accomplish this is to hande it manually,
iterating the default GMainContext inside the signal handler callback.
This is called a &quot;nested main loop&quot; and it&apos;s very dangerous because it
frequently results in code being executed at unexpected times. For
example, the attached bug report shows an unintended but also
unavoidable example of a triple nested main loop that occurs where
a second download starts inside the first decide-destination signal
handler, then another, then another. Not good. (Note this is not a new
problem in GTK 4. gtk_dialog_run() no longer exists because it
internally ran a nested main loop, which had this exact same problem.)

To fix this, we need to allow applications to handle the
decide-destination signal asynchronously, so they can finish the signal
handler without immediately providing a destination, provide it later on
in a future main context iteration, and know that the download will not
actually begin until it has been provided. This is easy to do because
the download implementation is already internally asynchronous. A basic
test is added to make sure it even works.

This commit also contains a few drive-by changes in TestDownloads.cpp,
notably where two of the decide-destination callbacks were missing
return values and therefore triggering undefined behavior. It also fixes
a typo that got copy/pasted around and removes some references to URIs
that are now paths
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(_WebKitDownloadPrivate::~_WebKitDownloadPrivate):
(maybeFinishDecideDestination):
(webkitDownloadDecideDestination):
(webkit_download_class_init):
(webkitDownloadDecideDestinationWithSuggestedFilename):
(webkit_download_set_destination):
(webkit_download_cancel):
* Source/WebKit/UIProcess/API/glib/WebKitDownloadClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitDownloadPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(downloadLocalFileSuccessfully):
(testDownloadOverwriteDestinationDisallowed):
(testDownloadLocalFileError):
(testDownloadRemoteFile):
(testDownloadRemoteFileError):
(testDownloadAsyncDecideDestination):
(testDownloadAsyncDecideDestinationCancel):
(testDownloadMIMEType):
(testDownloadTextPlainMIMEType):
(testDownloadUserAgent):
(testDownloadEphemeralContext):
(testDownloadDestinationURI):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/261118@main">https://commits.webkit.org/261118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/449d64677d91fd566f5c9f42ad3a03e35422dfcc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110584 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19664 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1937 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10793 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102856 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116326 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15704 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43982 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85892 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31900 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8857 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18240 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51516 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14760 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4197 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -73,12 +73,21 @@ enum {
+        if (decideDestinationCallback) {
+            g_critical("Bug: application handled WebKitDownload::decide-destination but failed to call webkit_download_set_destination() before the WebKitDownload was destroyed");
+            (*decideDestinationCallback)(AllowOverwrite::No, { });
+        }
```
CompletionHandler already asserts when destroyed without being invoked, I'm not sure we need this.

mcatanzaro replied:
Well it's optional to check this here, but (a) a g_critical() matches the precondition assertions we do using g_return_if_fail(), and (b) CompletionHandler asserts are only enabled in debug builds anyway, so app developers will never see them.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -73,12 +73,21 @@ enum {
+
     RefPtr<DownloadProxy> download;
 
+    std::optional<CompletionHandler<void(AllowOverwrite, String)>> decideDestinationCallback;
```
I don't think we need to use `std::optional` CompletionHandler can be set to nullptr

mcatanzaro replied:
Oops, didn't know.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -126,9 +135,23 @@ static void webkitDownloadGetProperty(GObject* object, guint propId, GValue* val
 
+static void maybeFinishDecideDestination(WebKitDownload* download)
+{
+    if (download->priv->decideDestinationCallback) {
```
if (auto completionHandler = std::exchange(download->priv->decideDestinationCallback, nullptr)) {

mcatanzaro replied:
Believe it or not, I actually wrote this at first because I've seen you do it several times and it seemed clever. Then I got rid of it to reduce horizontal scrolling. :P

mcatanzaro replied:
Anyway, I'll change this as you suggest.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -126,9 +135,23 @@ static void webkitDownloadGetProperty(GObject* object, guint propId, GValue* val
+        GUniquePtr<char> destinationPath(g_filename_from_uri(download->priv->destination.data(), nullptr, nullptr));
+        const char* destination = destinationPath.get();
+#endif
+        (*download->priv->decideDestinationCallback)(download->priv->allowOverwrite ? AllowOverwrite::Yes : AllowOverwrite::No, destination ? String::fromUTF8(destination) : emptyString());
```
String::fromUTF8 will return an empty string if destination is nullptr, I think. But this is not what we want, we say we use the suggested filename to sabe it under the downloads dir when not handled. If this was handled, then we should have a destination.

mcatanzaro replied:
Here we don't know whether it was handled or not. This code runs in both conditions.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -588,17 +588,30 @@ class WebViewDownloadTest: public WebViewTest {
+        GUniquePtr<char> destination(g_build_filename(Test::dataDirectory(), test->m_suggestedFilename.data(), nullptr));
+        webkit_download_set_destination(test->m_download.get(), destination.get());
+    }
+
```
Why is this static? it uses m_suggstedFilename and m_download

mcatanzaro replied:
There is a virtual member function with the same name. Probably too confusing. I'll just get rid of it.

mcatanzaro replied:
Decided to rename it, actually.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -588,17 +588,30 @@ class WebViewDownloadTest: public WebViewTest {
+        test->m_suggestedFilename = suggestedFilename;
+        if (test->m_shouldAsynchronouslyDecideDestination) {
+            g_idle_add(reinterpret_cast<GSourceFunc>(+[](WebViewDownloadTest* test) {
+                decideDestination(test);
```
Here you can do test->decideDestination()

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -588,17 +588,30 @@ class WebViewDownloadTest: public WebViewTest {
+                return G_SOURCE_REMOVE;
+            }), test);
+        } else
+            decideDestination(test);
```
And here too

carlosgcampos reviewed (approved):
Please add a test for cancel case too, it can be in a follow up if you prefer.

mcatanzaro committed:
[GLib] Need API for asynchronously handling WebKitDownload::decide-destination
https://bugs.webkit.org/show_bug.cgi?id=238748

Reviewed by Carlos Garcia Campos.

Currently WebKitDownload::decide-destination is emitted when WebKit
wants to know the filesystem location to download the file to. The
application is expected to display a file chooser using
gtk_dialog_run(), then return the result to WebKit using
webkit_download_set_destination(). In GTK 4, gtk_dialog_run() no longer
exists. The only way to accomplish this is to hande it manually,
iterating the default GMainContext inside the signal handler callback.
This is called a "nested main loop" and it's very dangerous because it
frequently results in code being executed at unexpected times. For
example, the attached bug report shows an unintended but also
unavoidable example of a triple nested main loop that occurs where
a second download starts inside the first decide-destination signal
handler, then another, then another. Not good. (Note this is not a new
problem in GTK 4. gtk_dialog_run() no longer exists because it
internally ran a nested main loop, which had this exact same problem.)

To fix this, we need to allow applications to handle the
decide-destination signal asynchronously, so they can finish the signal
handler without immediately providing a destination, provide it later on
in a future main context iteration, and know that the download will not
actually begin until it has been provided. This is easy to do because
the download implementation is already internally asynchronous. A basic
test is added to make sure it even works.

This commit also contains a few drive-by changes in TestDownloads.cpp,
notably where two of the decide-destination callbacks were missing
return values and therefore triggering undefined behavior. It also fixes
a typo that got copy/pasted around and removes some references to URIs
that are now paths
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(_WebKitDownloadPrivate::~_WebKitDownloadPrivate):
(maybeFinishDecideDestination):
(webkitDownloadDecideDestination):
(webkit_download_class_init):
(webkitDownloadDecideDestinationWithSuggestedFilename):
(webkit_download_set_destination):
(webkit_download_cancel):
* Source/WebKit/UIProcess/API/glib/WebKitDownloadClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitDownloadPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(downloadLocalFileSuccessfully):
(testDownloadOverwriteDestinationDisallowed):
(testDownloadLocalFileError):
(testDownloadRemoteFile):
(testDownloadRemoteFileError):
(testDownloadAsyncDecideDestination):
(testDownloadAsyncDecideDestinationCancel):
(testDownloadMIMEType):
(testDownloadTextPlainMIMEType):
(testDownloadUserAgent):
(testDownloadEphemeralContext):
(testDownloadDestinationURI):
(beforeAll):

Canonical link: https://commits.webkit.org/261118@main
